File: app/src/main/java/com/platdmit/simplecloudmanager/domain/converters/implement/LoadAverageConvertImp.java
package com.platdmit.simplecloudmanager.domain.converters.implement;

import com.platdmit.simplecloudmanager.data.api.models.ApiLoadAverage;
import com.platdmit.simplecloudmanager.data.database.entity.DbLoadAverage;
import com.platdmit.simplecloudmanager.domain.converters.LoadAverageConverter;
import com.platdmit.simplecloudmanager.domain.models.LoadAverage;

import java.util.ArrayList;
import java.util.List;

public class LoadAverageConvertImp implements LoadAverageConverter {
    private static final String CPU_NAME = "Процессор";
    private static final String RAM_NAME = "Память";
    private static final String ROM_NAME = "Дисковое хранилище";
    private static final String VAL_PERCENT = "%";
    private static final String VAL_SIZE = "Gb";

    @Override
    public List<DbLoadAverage> fromApiToDb(ApiLoadAverage apiLoadAverage, long serverId) {
        List<DbLoadAverage> loadAverageList = new ArrayList<>();
        loadAverageList.add(new DbLoadAverage(
                CPU_NAME,
                serverId,
                apiLoadAverage.cPU.getTotal(),
                apiLoadAverage.cPU.getHourly() + VAL_PERCENT,
                apiLoadAverage.cPU.getHourly()
        ));

        loadAverageList.add(new DbLoadAverage(
                RAM_NAME,
                serverId,
                apiLoadAverage.rAM.getTotal(),
                apiLoadAverage.rAM.getHourly() + VAL_PERCENT,
                apiLoadAverage.rAM.getHourly()
        ));

        loadAverageList.add(new DbLoadAverage(
                ROM_NAME,
                serverId,
                apiLoadAverage.disk.getTotal(),
                apiLoadAverage.disk.getFree() + VAL_SIZE,
                getFreePercent(apiLoadAverage.disk.getTotal(), apiLoadAverage.disk.getFree())
        ));

        return loadAverageList;
    }

    @Override
    public List<LoadAverage> fromApiToDomain(ApiLoadAverage apiLoadAverage) {
        List<LoadAverage> loadAverageList = new ArrayList<>();
        loadAverageList.add(new LoadAverage(
                CPU_NAME,
                apiLoadAverage.cPU.getTotal(),
                apiLoadAverage.cPU.getHourly() + VAL_PERCENT,
                apiLoadAverage.cPU.getHourly()
        ));

        loadAverageList.add(new LoadAverage(
                RAM_NAME,
                apiLoadAverage.rAM.getTotal(),
                apiLoadAverage.rAM.getHourly() + VAL_PERCENT,
                apiLoadAverage.rAM.getHourly()
        ));

        loadAverageList.add(new LoadAverage(
                ROM_NAME,
                apiLoadAverage.disk.getTotal(),
                apiLoadAverage.disk.getFree() + VAL_SIZE,
                getFreePercent(apiLoadAverage.disk.getTotal(), apiLoadAverage.disk.getFree())
        ));
        return loadAverageList;
    }

    @Override
    public List<LoadAverage> fromDbToDomain(List<DbLoadAverage> dbLoadAverages) {
        List<LoadAverage> loadAverageList = new ArrayList<>();
        for (DbLoadAverage dbLoadAverage : dbLoadAverages){
            loadAverageList.add(new LoadAverage(
                    dbLoadAverage.getName(),
                    dbLoadAverage.getTotal(),
                    dbLoadAverage.getValue(),
                    dbLoadAverage.getPercent()
            ));
        }
        return loadAverageList;
    }


    private double getFreePercent(String total, String free) {
        try {
            double numTotal = Double.parseDouble(total);
            double numFree = Double.parseDouble(free);
            return numFree / numTotal * 100;
        } catch (Exception e) {
            return 0;
        }
    }
}

File: domain/src/main/java/com/platdmit/domain/converters/implement/LoadAverageConvertImp.kt
package com.platdmit.domain.converters.implement

import com.platdmit.data.api.models.ApiLoadAverage
import com.platdmit.data.database.entity.DbLoadAverage
import com.platdmit.domain.converters.LoadAverageConverter
import com.platdmit.domain.models.LoadAverage
import java.util.*

class LoadAverageConvertImp : LoadAverageConverter {
    override fun fromApiToDb(apiLoadAverage: ApiLoadAverage, serverId: Long): List<DbLoadAverage> {
        val loadAverageList: MutableList<DbLoadAverage> = emptyList<DbLoadAverage>().toMutableList();
        loadAverageList.add(DbLoadAverage(
                CPU_NAME,
                serverId,
                apiLoadAverage.cPU.getTotal(), apiLoadAverage.cPU.getHourly().toString() + VAL_PERCENT,
                apiLoadAverage.cPU.getHourly()
        ))
        loadAverageList.add(DbLoadAverage(
                RAM_NAME,
                serverId,
                apiLoadAverage.rAM.getTotal(), apiLoadAverage.rAM.getHourly().toString() + VAL_PERCENT,
                apiLoadAverage.rAM.getHourly()
        ))
        loadAverageList.add(DbLoadAverage(
                ROM_NAME,
                serverId,
                apiLoadAverage.disk.getTotal(),
                apiLoadAverage.disk.getFree() + VAL_SIZE,
                getFreePercent(apiLoadAverage.disk.getTotal(), apiLoadAverage.disk.getFree())
        ))
        return loadAverageList
    }

    override fun fromApiToDomain(apiLoadAverage: ApiLoadAverage): List<LoadAverage> {
        val loadAverageList: MutableList<LoadAverage> = emptyList<LoadAverage>().toMutableList();
        loadAverageList.add(LoadAverage(
                CPU_NAME,
                apiLoadAverage.cPU.getTotal(), apiLoadAverage.cPU.getHourly().toString() + VAL_PERCENT,
                apiLoadAverage.cPU.getHourly()
        ))
        loadAverageList.add(LoadAverage(
                RAM_NAME,
                apiLoadAverage.rAM.getTotal(), apiLoadAverage.rAM.getHourly().toString() + VAL_PERCENT,
                apiLoadAverage.rAM.getHourly()
        ))
        loadAverageList.add(LoadAverage(
                ROM_NAME,
                apiLoadAverage.disk.getTotal(),
                apiLoadAverage.disk.getFree() + VAL_SIZE,
                getFreePercent(apiLoadAverage.disk.getTotal(), apiLoadAverage.disk.getFree())
        ))
        return loadAverageList
    }

    override fun fromDbToDomain(dbLoadAverages: List<DbLoadAverage>): List<LoadAverage> {
        return dbLoadAverages.map {
            LoadAverage(
                    it.name,
                    it.total,
                    it.value,
                    it.percent
            )
        }
    }

    private fun getFreePercent(total: String, free: String): Double {
        val numTotal = total.toDouble()
        val numFree = free.toDouble()
        return numFree / numTotal * 100
    }

    companion object {
        private const val CPU_NAME = "Процессор"
        private const val RAM_NAME = "Память"
        private const val ROM_NAME = "Дисковое хранилище"
        private const val VAL_PERCENT = "%"
        private const val VAL_SIZE = "Gb"
    }
}